cpp/src/fil/internal.cuh

```diff
@@ -266,6 +302,90 @@ struct forest_params_t {
 /// FIL_TPB is the number of threads per block to use with FIL kernels
 const int FIL_TPB = 256;
 
+static const uint32_t max_precise_int_float = 1 << 24;  // 16'777'216
+
+struct cat_feature {
+  int max_matching = -1, n_nodes = 0;
+  void reduce_with(cat_feature b)
+  {
+    max_matching = std::max(max_matching, b.max_matching);
+    n_nodes += b.n_nodes;
+  }
+};
+typedef std::vector<cat_feature> v_cat_feature;
+struct categorical_branches {
+  // arrays from each node ID are concatenated first, then from all categories
+  uint8_t* bits = nullptr;
+  // largest matching category in the model, per feature ID
+  int* max_matching = nullptr;
+  size_t bits_size = 0, max_matching_size = 0;
+
+  // set count is due to tree_idx + node_within_tree_idx are both ints, hence uint32_t result
+  template <typename node_t>
+  __host__ __device__ __forceinline__ int get_child(const node_t& node,
+                                                    int node_idx,
+                                                    float val) const
+  {
+    bool cond;
+    if (isnan(val)) {
+      cond = !node.def_left();
+    } else {
+      if (node.is_categorical()) {  // if the bit is not enabled, this is const false
+        int category = val;
+        // standard boolean packing. This layout has better ILP
+        // node.set() is global across feature IDs and is an offset (as opposed
+        // to set number). If we run out of uint32_t and we have hundreds of
+        // features with similar categorical feature count, we may consider
+        // storing node ID within nodes with same feature ID and look up
+        // {.max_matching, .first_node_offset} = ...[feature_id]
+        cond = (category <= max_matching[node.fid()]) &&
+               bits[node.set() + category / 8] & (1 << category % 8);
+      } else {
+        cond = val >= node.thresh();
+      }
+    }
+    return node.left(node_idx) + cond;
+  }
+  static int sizeof_mask_from_max_matching(int max_matching)
+  {
+    return raft::ceildiv(max_matching + 1, 8);
+  }
+  int sizeof_mask(int feature_id) const
+  {
+    return sizeof_mask_from_max_matching(max_matching[feature_id]);
+  }
+
+  // NB! no __device__ here
+  void host_allocate(v_cat_feature cf)
```

Comment:
I'd split this structure into two:

- a structure that owns the category bitmaps, and where the bitmaps are stored e.g. using `std::vector`, and a method to get the accessor;
- an accessor structure that only refers to the bitmap data (this time, through pointers).